source/fuzz/transformation_wrap_vector_synonym.cpp

```diff
@@ -0,0 +1,172 @@
+    uint32_t instruction_id, uint32_t vector_operand1, uint32_t vector_operand2,
+    uint32_t vec_id, uint32_t pos) {
+  message_.set_instruction_id(instruction_id);
+  message_.set_vector_operand1(vector_operand1);
+  message_.set_vector_operand2(vector_operand2);
+  message_.set_vec_id(vec_id);
+  message_.set_scalar_position(pos);
+}
+
+bool TransformationWrapVectorSynonym::IsApplicable(
+    opt::IRContext* ir_context,
+    const TransformationContext& transformation_context) const {
+  auto instruction =
+      ir_context->get_def_use_mgr()->GetDef(message_.instruction_id());
+  auto vec1 = ir_context->get_def_use_mgr()->GetDef(message_.vector_operand1());
+  auto vec2 = ir_context->get_def_use_mgr()->GetDef(message_.vector_operand2());
+
+  // |instruction_id| must refer to an existing instruction.
+  if (instruction == nullptr) {
+    return false;
+  }
+
+  std::unordered_set<SpvOp> valid_opcodes{SpvOpIAdd, SpvOpISub, SpvOpIMul,
+                                          SpvOpSDiv, SpvOpUDiv, SpvOpFAdd,
+                                          SpvOpFSub, SpvOpFMul, SpvOpFDiv};
+
+  if (!valid_opcodes.count(instruction->opcode())) {
+    return false;
+  }
+
+  // |vector_operand1| and |vector_operand2| must exist.
+  if (vec1 == nullptr || vec2 == nullptr) {
+    return false;
+  }
+
+  // The 2 vectors must be the same valid vector type.
+  auto vec1_type_id = vec1->type_id();
+  auto vec2_type_id = vec2->type_id();
+  auto vec_type = ir_context->get_type_mgr()->GetType(vec1_type_id);
+
+  if (vec1_type_id != vec2_type_id) {
+    return false;
+  }
+
+  if (vec_type->AsVector() == nullptr) {
+    return false;
+  }
+
+  // |vec_id| must be fresh.
+  if (!fuzzerutil::IsFreshId(ir_context, message_.vec_id())) {
+    return false;
+  }
+
+  std::vector<uint32_t> vec_ids{message_.vector_operand1(),
+                                message_.vector_operand2(), message_.vec_id()};
+
+  // |vec_id|, |result_id1| and |result_id2| should be disparate ids.
+  if (fuzzerutil::HasDuplicates(vec_ids)) {
+    return false;
+  }
+
+  // |scalar_position| needs to be a non-negative integer less than the vector
+  // length.
+  // OpTypeVector instruction has the component count at index 2.
+  auto vec_len = ir_context->get_def_use_mgr()
+                     ->GetDef(vec1_type_id)
+                     ->GetSingleWordOperand(2);
+  if (message_.scalar_position() >= vec_len) {
+    return false;
+  }
+
+  // Check the id in the corresponding position from the result vectors are
+  // synonyms with the operands from the original instruction.
+  uint32_t pos = message_.scalar_position();
+  uint32_t inst_operand1 = instruction->GetSingleWordInOperand(0);
+  uint32_t inst_operand2 = instruction->GetSingleWordInOperand(1);
+  bool check_synonym_operand1 =
+      transformation_context.GetFactManager()->IsSynonymous(
+          MakeDataDescriptor(message_.vector_operand1(), {pos}),
+          MakeDataDescriptor(inst_operand1, {}));
```

Comment:
Avoid introducing unnecessary temporary variables. Change this to:

```
if (!transformation_context.GetFactMAnager()->IsSynonymous(...)) {
   return false;
}
```

and the same for operand 2.

Reply:
Done.